Issue-Label-Bot-v2/notebooks/Repo_MLP.ipynb

```diff
@@ -0,0 +1,230 @@
+    "        self.download_embeddings_from_gcs()\n",
+    "        with open(self.embeddings_file, 'rb') as f:\n",
+    "            data = dpickle.load(f)\n",
+    "\n",
+    "        # filter labels\n",
+    "        c = Counter()\n",
+    "        for lbls in data['labels']:\n",
+    "            c.update(lbls)\n",
+    "        self.all_labels = [x for x in c if c[x] >= self.min_freq]\n",
+    "\n",
+    "        X = []\n",
+    "        y = []\n",
+    "        for emb, lbls in zip(data['features'], data['labels']):\n",
+    "            mask = [self.all_labels.index(x) for x in lbls if c[x] >= self.min_freq]\n",
+    "            if mask == []:\n",
+    "                continue\n",
+    "            zer = np.zeros(len(self.all_labels))\n",
+    "            zer[mask] = 1\n",
+    "            y.append(zer)\n",
+    "            X.append(emb)\n",
+    "        return X, y\n",
+    "\n",
+    "    def train(self):\n",
+    "        X, y = self.load_training_data()\n",
+    "        self.clf = MLPWrapper()\n",
+    "        # get probability thresholds before `fit` because it overwrites classifier\n",
+    "        self.clf.find_probability_thresholds(X, y)\n",
+    "        self.probability_thresholds = self.clf.probability_thresholds\n",
+    "        # train model using the whole data\n",
+    "        self.clf.fit(X, y)\n",
+    "        self.save_model()\n",
+    "\n",
+    "    def save_model(self):\n",
+    "        self.clf.save_model(model_file=self.model_file)\n",
+    "        # dump label columns for prediction\n",
+    "        label_dict = {'labels': self.all_labels, 'probability_thresholds': self.probability_thresholds}\n",
+    "        with open(self.labels_file, 'wb') as f:\n",
+    "            dpickle.dump(label_dict, f)\n",
+    "\n",
+    "        self.upload_model_to_gcs()\n",
+    "\n",
+    "    def upload_model_to_gcs(self):\n",
+    "        # upload model\n",
+    "        storage_client = storage.Client()\n",
+    "        bucket = storage_client.get_bucket(self.model_bucket_name)\n",
+    "        blob = bucket.blob(self.model_dest)\n",
+    "        blob.upload_from_filename(self.model_file)\n",
+    "\n",
+    "        # upload label columns\n",
+    "        storage_client = storage.Client()\n",
+    "        bucket = storage_client.get_bucket(self.model_bucket_name)\n",
+    "        blob = bucket.blob(self.labels_dest)\n",
+    "        blob.upload_from_filename(self.labels_file)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Run locally to test the code"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "r = RepoMLP()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "r.train()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create entry point using fairing"
+   ]
```

Comment:
Do you mind explaining what the below cell is doing as an explanation in the notebook?  This would be great for anyone who is browsing Kubeflow and wants to understand what is happening looking at this notebook.   Also would be good to explain what the fairing markers are doing in earlier cells and what functionality they provide

Reply:
I dug around for documentation on Fairing and it does not exist or is extremely sparse, so anything you can add here in terms of explanation and how it is working will be extremely helpful!

Reply:
Sure, I will do that